main/src/core/MonoDevelop.Ide/MonoDevelop.Components.Commands/CommandManager.cs

```diff
@@ -567,14 +591,182 @@ internal bool ProcessKeyEvent (Gdk.EventKey ev)
+							var isFirstView = IsFirstView (expectedKeyView, addViewBeforeChildren: true, removeContentView: true);
+							if (isFirstView) {
+								//if we are in first item, we change to the previous Gtk widget
+								backToGtk = true;
+							}
+						} else if ((int) currentEvent.ModifierFlags == (int) Mac.KeyModifierFlag.None) {
+							var isLastView = IsLastView (expectedKeyView, addViewBeforeChildren: true, removeContentView: true);
+							if (isLastView) {
+								//if we are in last element we change to next Gtk widget
+								backToGtk = true;
+							}
+						}
+					}
+
+					//source editor doesn't allow to change focus (we need improve this)
+					if (!backToGtk) {
+						SimulateKeyDownInView (view, expectedKeyView, gtkNSViewHost, currentEvent, window);
+					} else {
+
+						if (!gtkNSViewHost.IsFocus)
+							gtkNSViewHost.IsFocus = true;
+
+						//Unfocus all focused views in the Gtkquartzwindow
+						//NOTE: Be careful! after this, native window.FirstResponder it will change from the current one
+						window.MakeFirstResponder (null);
+					
+						//gtkNSViewHost.ProcessEvent (ev);
+						//GetGtkWindow (ev.Window).Present ();
+					}
+
+					//return true brokes the current event handling from Gtk, then if we back to gtk we should return false
+					return !backToGtk;
+				}
 			}
 #endif
 			return false;
 		}
 
+		const string sourceEditorTypeClass = ("Microsoft.VisualStudio.Text.Editor.Implementation.CocoaTextViewControl");
+		static bool IsSourceEditor (AppKit.NSView view) => view.GetType ().FullName == sourceEditorTypeClass;
+
+		internal List<AppKit.NSView> GetOrderedFocusableViews (AppKit.NSView view, bool addViewBeforeChildren, bool removeContentView)
+		{
+			var mainView = GetGtkNSViewHostContentView (view);
+			var focusableViews = new List<AppKit.NSView> ();
+			AddRecursivelyFocusableViews (mainView, focusableViews, addViewBeforeChildren, removeContentView ? mainView : null);
+			return focusableViews;
+		}
+
+		AppKit.NSView GetNextKeyView (AppKit.NSView view, bool addViewBeforeChildren, bool removeContentView)
+		{
+			var views = GetOrderedFocusableViews (view, addViewBeforeChildren, removeContentView);
+			var currentIndex = views.IndexOf (view);
+			if (currentIndex > -1 && currentIndex < views.Count - 1) {
+				return views [currentIndex + 1];
+			}
+			return null;
+		}
+		AppKit.NSView GetPreviousKeyView (AppKit.NSView view, bool addViewBeforeChildren, bool removeContentView)
+		{
+			var views = GetOrderedFocusableViews (view, addViewBeforeChildren, removeContentView);
+			var currentIndex = views.IndexOf (view);
+			if (currentIndex > 0 && currentIndex < views.Count) {
+				return views [currentIndex - 1];
+			}
+			return null;
+		}
+
+		bool IsLastView (AppKit.NSView view, bool addViewBeforeChildren, bool removeContentView) => GetOrderedFocusableViews (view, addViewBeforeChildren, removeContentView).LastOrDefault () == view;
```

Comment:
newline before implementation.
```suggestion
		bool IsLastView (AppKit.NSView view, bool addViewBeforeChildren, bool removeContentView)
			=> GetOrderedFocusableViews (view, addViewBeforeChildren, removeContentView).LastOrDefault () == view;
```